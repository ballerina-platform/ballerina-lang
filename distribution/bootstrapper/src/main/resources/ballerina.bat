@echo off

REM ---------------------------------------------------------------------------
REM   Copyright (c) 2019, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
REM
REM   Licensed under the Apache License, Version 2.0 (the "License");
REM   you may not use this file except in compliance with the License.
REM   You may obtain a copy of the License at
REM
REM   http://www.apache.org/licenses/LICENSE-2.0
REM
REM   Unless required by applicable law or agreed to in writing, software
REM   distributed under the License is distributed on an "AS IS" BASIS,
REM   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
REM   See the License for the specific language governing permissions and
REM   limitations under the License.

rem ---------------------------------------------------------------------------
rem Main Script for Ballerina
rem
rem Environment Variable Prerequisites
rem
rem   BALLERINA_HOME  Home of BALLERINA installation. If not set I will  try
rem                   to figure it out.
rem
rem   JAVA_HOME       Must point at your Java Development Kit installation.
rem
rem   JAVA_OPTS       (Optional) Java runtime options used when the commands
rem                   is executed.
rem ---------------------------------------------------------------------------

rem ----- if JAVA_HOME is not set we're not happy ------------------------------

:checkJava

set BALLERINA_HOME=%~sdp0..
if exist %BALLERINA_HOME%\..\..\dependencies\jdk8u265-b01-jre (
   set "JAVA_HOME=%BALLERINA_HOME%\..\..\dependencies\jdk8u265-b01-jre"
)

if "%JAVA_HOME%" == "" goto noJavaHome
if not exist "%JAVA_HOME%\bin\java.exe" goto noJavaHome
goto checkServer

:noJavaHome
echo "You must set the JAVA_HOME variable before running Ballerina."
goto end

rem ----- set BALLERINA_HOME ----------------------------
:checkServer
rem %~sdp0 is expanded pathname of the current script under NT with spaces in the path removed
set BALLERINA_HOME=%~sdp0..

goto updateClasspath

:noServerHome
echo BALLERINA_HOME is set incorrectly or BALLERINA could not be located. Please set BALLERINA_HOME.
goto end

rem ----- update classpath -----------------------------------------------------
:updateClasspath

setlocal EnableDelayedExpansion
set BALLERINA_CLASSPATH=
FOR %%C in ("%BALLERINA_HOME%\bre\lib\bootstrap\*.jar") DO set BALLERINA_CLASSPATH=!BALLERINA_CLASSPATH!;"%BALLERINA_HOME%\bre\lib\bootstrap\%%~nC%%~xC"

set BALLERINA_CLASSPATH="%JAVA_HOME%\lib\tools.jar";%BALLERINA_CLASSPATH%;

set BALLERINA_CLASSPATH=!BALLERINA_CLASSPATH!;"%BALLERINA_HOME%\bre\lib\*"

set BALLERINA_CLI_HEIGHT=
set BALLERINA_CLI_WIDTH=
for /F "tokens=2 delims=:" %%a in ('mode con') do for %%b in (%%a) do (
  if not defined BALLERINA_CLI_HEIGHT (
     set "BALLERINA_CLI_HEIGHT=%%b"
  ) else if not defined BALLERINA_CLI_WIDTH (
     set "BALLERINA_CLI_WIDTH=%%b"
  )
)

set argCount=0
for %%x in (%*) do (
   set /A argCount+=1
   set "argValue[!argCount!]=%%~x"
)

set /a counter=1
for /l %%i in (1, 1, %argCount%) do (
   set /a counter=!counter!+1
   if "!argValue[%%i]!"=="--debug" call set BAL_JAVA_DEBUG=%%!counter!
)

if defined BAL_JAVA_DEBUG goto commandDebug

rem ----- Process the input command -------------------------------------------
goto doneStart

rem ----- commandDebug ---------------------------------------------------------
:commandDebug
if "%BAL_JAVA_DEBUG%"=="" goto noDebugPort
if not "%JAVA_OPTS%"=="" echo Warning !!!. User specified JAVA_OPTS will be ignored, once you give the BAL_JAVA_DEBUG variable.
set JAVA_OPTS=-Xdebug -Xnoagent -Djava.compiler=NONE -Xrunjdwp:transport=dt_socket,server=y,suspend=y,address=%BAL_JAVA_DEBUG%
echo Please start the remote debugging client to continue...
goto runServer

:noDebugPort
echo Please specify the debug port for the BAL_JAVA_DEBUG variable
goto end

:doneStart
if "%OS%"=="Windows_NT" @setlocal
if "%OS%"=="WINNT" @setlocal

goto runServer

rem ----------------- Execute The Requested Command ----------------------------

:runServer

set CMD=%*

rem ---------- Add jars to classpath ----------------

set BALLERINA_CLASSPATH=.\bre\lib\bootstrap;%BALLERINA_CLASSPATH%

rem BALLERINA_CLASSPATH_EXT is for outsiders to additionally add
rem classpath locations, e.g. AWS Lambda function libraries.
set BALLERINA_CLASSPATH=%BALLERINA_CLASSPATH%;%BALLERINA_CLASSPATH_EXT%

set CMD_LINE_ARGS=-Xbootclasspath/a:%BALLERINA_XBOOTCLASSPATH% -Xms256m -Xmx1024m -XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath="%BALLERINA_HOME%\heap-dump.hprof"  -Dcom.sun.management.jmxremote -classpath %BALLERINA_CLASSPATH% %JAVA_OPTS% -Dballerina.home="%BALLERINA_HOME%" -Dballerina.target="jvm" -Djava.command="%JAVA_HOME%\bin\java" -Djava.opts="%JAVA_OPTS%" -Denable.nonblocking=false -Dfile.encoding=UTF8 -Dballerina.version=1.0.3 -Djava.util.logging.config.class="org.ballerinalang.logging.util.LogConfigReader" -Djava.util.logging.manager="org.ballerinalang.logging.BLogManager" -DBALLERINA_DEV_COMPILE_BALLERINA_ORG=true

set jar=%2
if "%1" == "run" if not "%2" == "" if "%jar:~-4%" == ".jar" goto runJarFile
:runJava
"%JAVA_HOME%\bin\java" %CMD_LINE_ARGS% org.ballerinalang.tool.Main %CMD%
goto end

:runJarFile
for /f "tokens=1,*" %%a in ("%*") do set ARGS=%%b
"%JAVA_HOME%\bin\java" %CMD_LINE_ARGS% -jar %ARGS%
goto end

:end
goto endlocal

:endlocal

:END
