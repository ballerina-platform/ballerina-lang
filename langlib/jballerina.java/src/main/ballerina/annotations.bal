// Copyright (c) 2019 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
//
// WSO2 Inc. licenses this file to you under the Apache License,
// Version 2.0 (the "License"); you may not use this file except
// in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

# Represents a Java class name.
public type Class string;

# Represents a Java identifier.
public type Identifier string;

# Represents a Java array type. It is used to specify the parameter types in the `java:Constructor` and `java:Method`
# annotations.
#
# + class - Element class of the array type
# + dimensions - Dimensions of the array type
public type ArrayType record {|
    Class 'class;
    byte dimensions;
|};

# Describes a Java constructor. If the `paramTypes` field is not specified, then the parameter types are inferred from
# the corresponding Ballerina function.
#
# + class - The class in which the constructor exists
# + paramTypes - An optional field, which describes the parameter types of the constructor
public type ConstructorData record {|
    Class 'class;
    (Class | ArrayType)[] paramTypes?;
|};

# Describes a Java method. If the `paramTypes` field is not specified, then the parameter types are inferred from the
# corresponding Ballerina function.
#
# + name - An optional field, which describes the name of the Java method. If this field is not provided, then the name
#          is inferred from the Ballerina function name
# + class - The class in which the method exists
# + paramTypes - An optional field, which describes the parameter types of the method
public type MethodData record {|
    Identifier name?;
    Class 'class;
    (Class | ArrayType)[] paramTypes?;
|};

# Describes a Java field.
#
# + name - An optional field, which describes the name of the Java field. If this field is not provided,
#          then the name is inferred from the Ballerina function name
# + class - The class in which the field exists
public type FieldData record {|
    Identifier name;
    Class 'class;
|};

# Describes a Java class that corresponds to a Ballerina object.
#
# + class - The Java class represented by the Ballerina object.
public type ObjectData record {|
    Class 'class;
|};

# Describes a Java constructor, which provides an implementation of a Ballerina function of which the body is marked as
# `external`. If the Ballerina function body is marked as `external`, it means that the implementation of the
# function is not provided in the Ballerina source module.
#
# The following code snippet shows an example usage of this annotation. Here, the `newJavaLinkedList` Ballerina function's
# implementation is provided by the default constructor of the `java.util.LinkedList` class.
# ```ballerina
# function newJavaLinkedList() returns handle = @java:Constructor {
#      'class: "java.util.LinkedList"
# } external;
# ```
public const annotation ConstructorData Constructor on source external;

# Describes a Java method, which provides an implementation of a Ballerina function of which the body is marked as
# `external`. If the Ballerina function body is marked as `external`, it means that the implementation of the
# function is not provided in the Ballerina source module.
#
# The following code snippet shows an example usage of this annotation. Here, the `getUUID` Ballerina function's
# implementation is provided by the `java.util.UUID.randomUUID` static method.
# ```ballerina
# function getUUID() returns handle = @java:Method {
#     name: "randomUUID",
#     'class: "java.util.UUID"
# } external;
# ```
# The `name` field is optional. If it is not provided, the name of the Java method is inferred
# from the Ballerina function.
public const annotation MethodData Method on source external;

# Describes a Java Field access, which provides an implementation of a Ballerina function of which the body is marked as
# `external`.
# ```ballerina
# function getError() returns handle = @java:FieldGet {
#     name:"err",
#     'class:"java/lang/System"
# } external;
# ```
public const annotation FieldData FieldGet on source external;

# Describes a Java Field mutate, which provides an implementation of a Ballerina function of which the body is marked as
# `external`.
# ```ballerina
# function setContractId(handle contractId) = @java:FieldSet {
#   name:"contractId",
#   'class:"org/lang/impl/JavaFieldAccessMutate"
# } external;
# ```
public const annotation FieldData FieldSet on source external;

# Describes the Java class representing a Ballerina binding.
# ```ballerina
# @java:Binding {
#   'class: "java.io.File"
# }
# ```
public const annotation ObjectData Binding on class;
