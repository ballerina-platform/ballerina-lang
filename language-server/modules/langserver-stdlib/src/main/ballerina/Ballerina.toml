[package]
org = "ballerina"
name = "module1"
version = "0.1.0"

[[platform.java11.dependency]]
path = "../libs/language-server-stdlib-${version}.jar"
groupId = "ballerina"
artifactId = "mock"
version = "0.0.0"
