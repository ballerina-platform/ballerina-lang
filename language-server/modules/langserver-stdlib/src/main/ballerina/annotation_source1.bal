
public type AnnotationType record {
    string foo;
    int bar?;
};

public annotation AnnotationType annotationCommon;

public annotation AnnotationType a1 on service;

public annotation AnnotationType a2 on function;

annotation AnnotationType a3 on service;

public annotation AnnotationType typeAnnotation1 on type;

public const annotation typeAnnotation2 on type;

public annotation typeAnnotation3 on type;

public annotation AnnotationType classAnnotation1 on class;

public annotation classAnnotation2 on class;

public const annotation classAnnotation3 on source class;

public annotation AnnotationType functionAnnotation1 on function;

public annotation functionAnnotation2 on function;

public annotation AnnotationType objectFunctionAnnotation1 on object function;

public annotation objectFunctionAnnotation2 on object function;

public annotation AnnotationType parameterAnnotation1 on parameter;

public annotation parameterAnnotation2 on parameter;

public annotation AnnotationType returnAnnotation1 on return;

public annotation returnAnnotation2 on return;

public annotation AnnotationType serviceAnnotation1 on service;

public annotation serviceAnnotation2 on service;

public annotation AnnotationType fieldAnnotation1 on field;

public annotation fieldAnnotation2 on field;

public annotation AnnotationType objectFieldAnnotation1 on object field;

public annotation objectFieldAnnotation2 on object field;

public annotation AnnotationType recordFieldAnnotation1 on record field;

public annotation recordFieldAnnotation2 on record field;

public const annotation AnnotationType sourceAnnotationAnnotation1 on source annotation;

public const annotation sourceAnnotationAnnotation2 on source annotation;

public const annotation AnnotationType sourceExternalAnnotation1 on source external;

public const annotation sourceExternalAnnotation2 on source external;

public const annotation AnnotationType sourceVarAnnotation1 on source var;

public const annotation sourceVarAnnotation2 on source var;

public const annotation AnnotationType sourceConstAnnotation1 on source const;

public const annotation sourceConstAnnotation2 on source const;

public const annotation AnnotationType sourceListenerAnnotation1 on source listener;

public const annotation sourceListenerAnnotation2 on source listener;

public const annotation AnnotationType sourceWorkerAnnotation1 on source worker;

public const annotation sourceWorkerAnnotation2 on source worker;

public const annotation commonAnnotation1;

public const annotation AnnotationType commonAnnotation2;

public const annotation AnnotationType serviceRemoteFunctionAnnotation1 on service remote function;

public const annotation serviceRemoteFunctionAnnotation2 on service remote function;

public const annotation AnnotationType sourceClientDeclaration1 on source client;

public const annotation sourceClientDeclaration2 on source client;
