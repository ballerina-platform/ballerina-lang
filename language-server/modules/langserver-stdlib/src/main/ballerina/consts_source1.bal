
public const int TEST_INT_CONST1 = 1;

public const string TEST_STRING_CONST1 = "HELLO WORLD";

const int TEST_INT_CONST2 = 2;

const string TEST_STRING_CONST2 = "HELLO WORLD2";

public future<int> TEST_FUTURE_INT = getFutureInt();

function getInt() returns int {
    return 1;
}

function getFutureInt() returns future<int> {
    return start getInt();
}