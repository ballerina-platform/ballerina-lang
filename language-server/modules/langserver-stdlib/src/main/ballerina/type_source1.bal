
public type TestRecord1 record {
    int rec1Field1 = 12;
    string rec1Field2 = "TestRecord1";
    int optionalInt?;
};

public type TestRecord2 record {
    int rec2Field1 = 12;
    string rec3Field2 = "TestRecord2";
};

type TestRecord3 record {
    int rec3Field1 = 12;
    string rec3Field2 = "TestRecord3";
};

type TestMap1 map<int>;

public type TestMap2 map<string>;

public type TestMap3 map<int>;

public class TestClass1 {
	public int field1;
	int field2;
	public function init(int field1, int field2) {
	    self.field1 = field1;
	    self.field2 = field2;
    }
}

public type TestObject1 object {
    
};

public type ErrorOne error<record { int code; }>;

public type ErrorTwo error<record { string[] alternatives; }>;

public type Error1 E1|E2;
public type E1 distinct error;
public type E2 distinct error;
public type E3 distinct error;
public type Error2 E3|int;
public type Error Error1|Error2;
