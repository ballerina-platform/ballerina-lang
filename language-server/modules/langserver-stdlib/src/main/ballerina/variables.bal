public int GLOBAL_VAR = 200;
