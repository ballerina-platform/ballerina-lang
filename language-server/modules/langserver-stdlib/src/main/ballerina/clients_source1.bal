import ballerina/jballerina.java;

# The HTTP client provides the capability for initiating contact with a remote HTTP service. The API it
# provides includes functions for the standard HTTP methods, forwarding a received request and sending requests
# using custom HTTP verbs.

# + url - Target service url
public client class Client {
    public string url;

    public function init(string url) {
        self.url = url;
    }

    # The `Client.post()` function can be used to send HTTP POST requests to HTTP endpoints.
    #
    # + path - Resource path
    # + message - An HTTP outbound request message or any payload of type `string`, `xml`, `json`, `byte[]`
    # + targetType - Specifies the target type
    # + return - The response for the request
    remote function post(@untainted string path, RequestMessage message, string targetType="xml") returns Response {
        Response response = new();
        return response;
    }
    
    # Sample resource method.
    #
    # + id1 - Path parameter
    # + ids - Rest path parameter
    # + str - Argument
    # + ids2 - Rest argument
    # + return - The response for the request
    resource function post path1/[string id1]/path2/[string... ids](string str, string... ids2) returns Response {
        return new Response();
    }
    
    # Sample remote method with java interoperability
    #
    # + path - Resource path
    # + targetType - Response or `anydata`, which is expected to be returned after data binding
    # + return - The Response or the ClientError
    remote function get(string path, TargetType targetType = <>) returns targetType|ClientError = @java:Method {
        'class: "org.ballerinalang.langserver.stdlib.ClientAction"
    } external;
    
    # Sample remote method with java interoperability
    #
    # + path - Resource path
    # + request - Request need to be forward
    # + targetType - Response or `anydata`, which is expected to be returned after data binding
    # + return - Response
    remote function forward(string path, handle request, TargetType targetType = <>) returns targetType = @java:Method {
        'class: "org.ballerinalang.langserver.stdlib.ClientAction"
    } external;
    
    # Sample remote method with java interoperability
    #
    # + targetType - `any`type which is expected to be returned after data binding
    # + return - Type any
    remote function delete(TargetType2 targetType = <>) returns targetType = @java:Method {
        'class: "org.ballerinalang.langserver.stdlib.ClientAction"
    } external;
    
    # Sample resource function with java interoperability
    #
    # + targetType - Response or `anydata`, which is expected to be returned after data binding
    # + return - The Response or the ClientError
    resource function get path3(TargetType targetType = <>) returns targetType|ClientError = @java:Method {
        'class: "org.ballerinalang.langserver.stdlib.ClientAction",
        name: "delete"
    } external;
    
    # Sample resource function with multiple target types with java interoperability
    #
    # + targetType - Response or `anydata`, which is expected to be returned after data binding
    # + return - The Response or the ClientError
    resource function get path4/[string pathParam](TargetType targetType = <>) returns targetType|ClientError|error = @java:Method {
        'class: "org.ballerinalang.langserver.stdlib.ClientAction",
        name: "get"
    } external;
    
    # Sample resource function with rest path praram with multiple target types with java interoperability
    #
    # + path - Request path
    # + message - An HTTP outbound request or any allowed payload
    # + headers - The entity headers
    # + mediaType - The MIME type header of the request entity
    # + targetType - HTTP response or `anydata`, which is expected to be returned after data binding
    # + return - The response or the payload (if the `targetType` is configured) or an `http:ClientError` if failed to
    # establish the communication with the upstream server or a data binding failure
    isolated resource function post [string ...path](RequestMessage message, map<string|string[]>? headers = (), string?
            mediaType = (), TargetType targetType = <>) returns targetType|ClientError = @java:Method {
        'class: "org.ballerinalang.langserver.stdlib.ClientAction",
        name: "postResource"
    } external;
}

# Represents a response.
#
# + statusCode - The response status code
# + reasonPhrase - The status code reason phrase
public class Response {
    public int statusCode = 200;
    public string reasonPhrase = "Test Reason phrase";
}

# The types of messages that are accepted by HTTP `listener` when sending out the outbound response.
public type ResponseMessage string|xml|json|byte[]|();

# The types of messages that are accepted by HTTP `client` when sending out the outbound request.
public type RequestMessage string|xml|json|byte[]|();

# The types of data values that are expected by the `client` to return after the data binding operation.
public type TargetType typedesc<Response|anydata>;

# Defines the possible client error types.
public type ClientError distinct error<map<anydata>>;

# The super type of all the types.
public type TargetType2 typedesc<any>;

