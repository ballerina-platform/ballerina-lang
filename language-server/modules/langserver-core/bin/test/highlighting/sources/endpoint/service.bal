import ballerina/http;
import ballerina/log;

service hello on new http:Listener(9090) {

    resource function sayHello(http:Caller caller, http:Request req) {
        var result = caller->respond("Hello, World!");

        http:Client myClient = new ("");

        if (result is error) {
            log:printError("Error sending response", result);
        }
    }
}
