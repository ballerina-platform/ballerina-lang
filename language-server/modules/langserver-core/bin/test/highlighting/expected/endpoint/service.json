{
  "source": {
    "file": "service.bal"
  },
  "result": [
    {
      "line": 5,
      "token": "WzQzLCA2LCAwXQ=="
    },{
      "line": 6,
      "token": "WzIxLCA2LCAwXQ=="
    },
    {
      "line":8,
      "token":"WzIwLCA4LCAwXQ=="
    }
  ]
}